crates/saya/core/src/verifier/starknet.rs

```diff
@@ -1,75 +1,81 @@
-use std::time::Duration;
-
 use anyhow::Context;
 use dojo_world::migration::TxnConfig;
 use dojo_world::utils::TransactionExt;
+use itertools::Itertools;
 use starknet::accounts::{Account, Call, ConnectedAccount};
-use starknet::core::types::{Felt, TransactionExecutionStatus, TransactionStatus};
+use starknet::core::types::Felt;
 use starknet::core::utils::get_selector_from_name;
-use starknet::providers::Provider;
-use tokio::time::sleep;
+use starknet_crypto::poseidon_hash_many;
+use tracing::trace;
 
-use crate::dojo_os::get_starknet_account;
-use crate::StarknetAccountData;
+use super::utils::wait_for_sent_transaction;
+use crate::SayaStarknetAccount;
 
 pub async fn starknet_verify(
     fact_registry_address: Felt,
     serialized_proof: Vec<Felt>,
-    starknet_config: StarknetAccountData,
+    cairo_version: Felt,
+    account: &SayaStarknetAccount,
 ) -> anyhow::Result<(String, Felt)> {
+    if serialized_proof.len() > 2000 {
+        trace!(
+            "Calldata too long at: {} felts, transaction could fail, splitting it.",
+            serialized_proof.len()
+        );
+    }
+
     let txn_config = TxnConfig { wait: true, receipt: true, ..Default::default() };
-    let account = get_starknet_account(starknet_config)?;
-    let account = account.lock().await;
+
+    let mut nonce = account.get_nonce().await?;
+    let mut hashes = Vec::new();
+
+    for fragment in serialized_proof.into_iter().chunks(2000).into_iter() {
```

Comment:
If a fragment upload fail, do we have a way that saya can make a call on the chain to resume from where it was?

Reply:
Not at the moment, but should be pretty straight forward to implement.

Reply:
Same then let's keep that in the track for a follow up PR to be implemented to ensure more reliable settlement.